### PR TITLE
cephfs_mirror: provide metrics for last successful snapshot sync

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/65171
Fixes: https://tracker.ceph.com/issues/64483
Fixes: https://tracker.ceph.com/issues/65629

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 1]
```diff
@@ -33,11 +33,15 @@ using namespace std;
 enum {
   l_cephfs_mirror_peer_replayer_first = 6000,
   l_cephfs_mirror_peer_replayer_snaps_synced,
+  l_cephfs_mirror_peer_replayer_sync_bytes,
```
Comment (vshankar) [inline src/tools/cephfs_mirror/PeerReplayer.cc:36]: why change the enumerated value for this?

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/tools/cephfs_mirror/PeerReplayer.cc (hunk 1 above)
Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc:36]: oops, I just made them in order. Didn't thought about the value change back then. I'll revert it if needed?

Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: @vshankar Fixed

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
rebased

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test api

**Review by vshankar — COMMENTED** [review]
Looks fine. IMO we should also add some tests for this change. WDYT @joscollin

**Comment by joscollin** [comment]
> Looks fine. IMO we should also add some tests for this change. WDYT @joscollin

@vshankar We could write tests for this change. Could you please clarify which commit changes needs tests? 
The `sync_bytes` or the metrics ?

**Comment by vshankar** [comment]
> > Looks fine. IMO we should also add some tests for this change. WDYT @joscollin
> 
> @vshankar We could write tests for this change. Could you please clarify which commit changes needs tests? The `sync_bytes` or the metrics ?

The perf counters added in this change - start/end/duration/sync_bytes

**Comment by joscollin** [comment]
> > > Looks fine. IMO we should also add some tests for this change. WDYT @joscollin
> > 
> > 
> > @vshankar We could write tests for this change. Could you please clarify which commit changes needs tests? The `sync_bytes` or the metrics ?
> 
> The perf counters added in this change - start/end/duration/sync_bytes

Let me check if the existing tests in test_mirroring.py could make use of it.

**Comment by joscollin** [comment]
> > > Looks fine. IMO we should also add some tests for this change. WDYT @joscollin
> > 
> > 
> > @vshankar We could write tests for this change. Could you please clarify which commit changes needs tests? The `sync_bytes` or the metrics ?
> 
> The perf counters added in this change - start/end/duration/sync_bytes

@vshankar 
I've checked the existing tests in test_mirroring.py once again. The existing tests  doesn't require checking starttime/endtime/duration/size of a snapshot sync to pass. There's no point in adding an additional time/size check to decide the test passed or failed. The existing tests basically checks if the sync is completed or not to decide a test pass/fail. So I think we will leave it as is and make use of these perf counters when appropriate need arises.

May be, we could write a new test that verifies if starttime/endtime/duration/size perf counters are populated and their existence in `counter dump`. But as we can see in test_mirroring.py, none of the `last_synced_snap` values are being checked for it's existence. @vshankar WDYT?

**Comment by vshankar** [comment]
> May be, we could write a new test that verifies if starttime/endtime/duration/size perf counters are populated and their existence in `counter dump`. But as we can see in test_mirroring.py, none of the `last_synced_snap` values are being checked for it's existence. @vshankar WDYT?

Correct - just a straightforward verification would suffice here.

**Comment by joscollin** [comment]
@vshankar 
Tests passed:
https://pulpito.ceph.com/jcollin-2024-07-28_00:45:57-fs:mirror-wip-jcollin-testing-distro-default-smithi/
https://pulpito.ceph.com/jcollin-2024-07-28_00:46:33-fs:valgrind-wip-jcollin-testing-distro-default-smithi/

**Comment by joscollin** [comment]
jenkins test api

File: qa/tasks/cephfs/test_mirroring.py [hunk 2]
```diff
@@ -567,10 +567,14 @@ def test_cephfs_mirror_stats(self):
...
 
-        # check snaps_synced
+        # check perf counters
         res = self.mirror_daemon_command(f'counter dump for fs: {self.primary_fs_name}', 'counter', 'dump')
         second = res[TestMirroring.PERF_COUNTER_KEY_NAME_CEPHFS_MIRROR_PEER][0]
         self.assertGreater(second["counters"]["snaps_synced"], first["counters"]["snaps_synced"])
+        self.assertGreater(second["counters"]["last_synced_start"], first["counters"]["last_synced_start"])
+        self.assertGreater(second["counters"]["last_synced_end"], second["counters"]["last_synced_start"])
+        self.assertGreater(second["counters"]["last_synced_duration"], 0)
+        self.assertGreater(second["counters"]["last_synced_bytes"], 0)
```
Comment (vshankar) [inline qa/tasks/cephfs/test_mirroring.py:577]: Ideally, you can compare this to the file size (or at least least greater than or equal to that considering some metadata amounted).

File: qa/tasks/cephfs/test_mirroring.py [hunk 3]
```diff
@@ -587,10 +591,14 @@ def test_cephfs_mirror_stats(self):
...
 
-        # check snaps_synced
+        # check perf counters
         res = self.mirror_daemon_command(f'counter dump for fs: {self.primary_fs_name}', 'counter', 'dump')
         third = res[TestMirroring.PERF_COUNTER_KEY_NAME_CEPHFS_MIRROR_PEER][0]
         self.assertGreater(third["counters"]["snaps_synced"], second["counters"]["snaps_synced"])
+        self.assertGreater(third["counters"]["last_synced_start"], second["counters"]["last_synced_end"])
+        self.assertGreater(third["counters"]["last_synced_end"], third["counters"]["last_synced_start"])
+        self.assertGreater(third["counters"]["last_synced_duration"], 0)
+        self.assertGreater(third["counters"]["last_synced_bytes"], 0)
```
Comment (vshankar) [inline qa/tasks/cephfs/test_mirroring.py:601]: ditto

File: qa/tasks/cephfs/test_mirroring.py (hunk 2 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_mirroring.py:577]: I'd like to do such kind of check, but `create_n_files` doesn't return any size info. We just have `50`. Is that 50 bytes total?

Comment (vshankar) [reply qa/tasks/cephfs/test_mirroring.py:577]: Oh I meant write some data to some files and then expect the perf counter to account that size.

File: doc/cephfs/cephfs-mirroring.rst [hunk 4]
```diff
@@ -398,6 +398,69 @@ created, the mirror daemon will clear the failed state upon successful synchroni
 When mirroring is disabled, the respective `fs mirror status` command for the file system
 will not show up in command help.
 
+Metrics
+-------
+
+CephFS exports mirroring metrics as :ref:`Labeled Perf Counters` which will be consumed by the OCP/ODF Dashboard to provide monitoring of Geo Replication in the OCP and ACM dashboard and elsewhere. This would generate the progress of cephfs_mirror syncing and thus provide the monitoring capability. CephFS exports the following mirroring metrics, which could be displayed using the ``counter dump`` command.
```
Comment (anthonyeleven) [inline doc/cephfs/cephfs-mirroring.rst:404]: Reword in present tense please.  And give some context for what these OCP and ACM dashboards are.

File: doc/cephfs/cephfs-mirroring.rst [hunk 5]
```diff
@@ -398,6 +398,69 @@ created, the mirror daemon will clear the failed state upon successful synchroni
...
+     - The total number of snapshots deleted
+   * - snaps_renamed
+     - Counter
+     - The total number of snapshots renamed
+   * - avg_sync_time
+     - Gauge
+     - The average time taken by all snapshot synchronizations
+   * - last_synced_start
+     - Gauge
+     - The sync start time of the Last Synced Snapshot
```
Comment (anthonyeleven) [inline doc/cephfs/cephfs-mirroring.rst:453]: This and the below: last synched snapshot lowercase

File: doc/cephfs/cephfs-mirroring.rst (hunk 4 above)
Comment (joscollin) [reply doc/cephfs/cephfs-mirroring.rst]: I don't remember from where I got (copy-pasted) that text. Updated the paragraph and dropped over-mentioning about the users. Please check now.

File: doc/cephfs/cephfs-mirroring.rst (hunk 5 above)
Comment (joscollin) [reply doc/cephfs/cephfs-mirroring.rst]: Updated.

File: qa/tasks/cephfs/test_mirroring.py (hunk 2 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_mirroring.py:577]: Got it, fixing it.

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
@vshankar Test passed: https://pulpito.ceph.com/jcollin-2024-07-29_11:06:40-fs:mirror-wip-jcollin-testing-distro-default-smithi/

**Comment by joscollin** [comment]
jenkins test make check

File: qa/tasks/cephfs/test_mirroring.py [hunk 6]
```diff
@@ -587,10 +592,14 @@ def test_cephfs_mirror_stats(self):
...
 
-        # check snaps_synced
+        # check perf counters
         res = self.mirror_daemon_command(f'counter dump for fs: {self.primary_fs_name}', 'counter', 'dump')
         third = res[TestMirroring.PERF_COUNTER_KEY_NAME_CEPHFS_MIRROR_PEER][0]
         self.assertGreater(third["counters"]["snaps_synced"], second["counters"]["snaps_synced"])
+        self.assertGreater(third["counters"]["last_synced_start"], second["counters"]["last_synced_end"])
+        self.assertGreater(third["counters"]["last_synced_end"], third["counters"]["last_synced_start"])
+        self.assertGreater(third["counters"]["last_synced_duration"], 0)
+        self.assertEquals(third["counters"]["last_synced_bytes"], 99614720)
```
Comment (vshankar) [inline qa/tasks/cephfs/test_mirroring.py:602]: nit: just put a comment on what this value is compared to.

**Review by vshankar — COMMENTED** [review]
Otherwise LGTM

**Review by vshankar — APPROVED** [review]

**Comment by joscollin** [comment]
Tests passed:
https://pulpito.ceph.com/jcollin-2024-07-29_15:03:32-fs:mirror-wip-jcollin-testing-distro-default-smithi/
https://pulpito.ceph.com/jcollin-2024-07-29_15:03:58-fs:valgrind-wip-jcollin-testing-distro-default-smithi/

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test make check arm64

**Comment by joscollin** [comment]
@vshankar Updated the doc

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67252](https://tracker.ceph.com/issues/67252).

**Comment by joscollin** [comment]
@vshankar Updated only the doc.

**Comment by joscollin** [comment]
@vshankar Updated only docs.

**Comment by vshankar** [comment]
> @vshankar Updated only docs.

kk, I already included a previous version in the branch, so as long as the latest push is doc only, I'm fine with that.

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240731064922-debug

**Pull request merged**